### PR TITLE
NMS-10586: Add RocksDB as queue for offheap messages

### PR DESCRIPTION
**Pull request by cgorantla**
This replaces H2 with RocksDB for OffHeap Queue.
RocksDB allows messages to be stored on file system.
Messages can be retrieved across minion restarts as well.

### External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-10586
* Bamboo (Continuous Integration): https://bamboo.opennms.org/

### CODE REVIEW

**Comment by j-white** [comment]
If we don't try and restore the messages after a restart, what happens to them? Can't we dispatch them anyway event if we no longer have a future reference to resolve after?

File: opennms-doc/guide-admin/src/asciidoc/text/minion/offheap.adoc [hunk 1]
```diff
@@ -29,6 +27,13 @@ echo 'offHeapSize=1.2GB
 enableOffHeap=true' > "$MINION_HOME/etc/org.opennms.core.ipc.sink.offheap.cfg"
 ----
 
+Also configure path where Offheap messages can be stored.
+
+[source, sh]
+----
+echo 'offHeapPath=/tmp/rocksdb' > "$MINION_HOME/etc/org.opennms.core.ipc.sink.offheap.cfg"
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply opennms-doc/guide-admin/src/asciidoc/text/minion/offheap.adoc]: Yeah, having this in Karaf data directory should be good enough. User shouldn't upgrade if they  still have Offheap messages to consume. Number of messages can be queried through JMX metrics.

File: core/ipc/sink/common/src/test/java/org/opennms/core/ipc/sink/offheap/AsyncDispatcherOffHeapTest.java [hunk 2]
```diff
@@ -125,109 +131,6 @@ public void testOffHeapStore() throws Exception {
         CompletableFuture.allOf(futures.toArray(new CompletableFuture<?>[]{}));
         asyncDispatcher.close();
     }
-    @Test(timeout=4*30*1000)
-    public void testRejectedWhenOffHeapIsFull() throws Exception {
```
[the remark this answers is not in the document]
Comment (cgorantla) [reply core/ipc/sink/common/src/test/java/org/opennms/core/ipc/sink/offheap/AsyncDispatcherOffHeapTest.java]: I was depending on current size of RocksDB which was not returning correctly. I fixed that.  Refactored these tests and they seem to work reliably now.

**Comment by cgorantla** [comment]
> If we don't try and restore the messages after a restart, what happens to them? Can't we dispatch them anyway event if we no longer have a future reference to resolve after?

True. We don't need future to dispatch messages. Fixed that

**Comment by cgorantla** [comment]
Looks like something is going wrong when build systems run RocksDB tests. I m not having any issue locally. I will try to debug this in circleci

**Comment by cgorantla** [comment]
Looks like bamboo/CircleCI encountered JVM crashes (due to JNI code from RocksDB) when it runs tests that should ideally throw exception for maximum space reached. Instead it always hangs in case of Circleci/Bamboo.   The difference with my local environment is the OS. I m using fedora where as CI/CD runs on Ubuntu. Going to debug more in Ubuntu environment.

**Comment by cgorantla** [comment]
Closing it for now. Will open once I actually start working on it again

**Pull request closed without merging**